api/src/main/java/org/openmrs/api/impl/ProgramWorkflowServiceImpl.java

```diff
@@ -74,35 +74,45 @@ public Program saveProgram(Program program) throws APIException {
 			throw new APIException("Program.concept.required", (Object[]) null);
 		}
 		
-		// ProgramWorkflow
 		for (ProgramWorkflow workflow : program.getAllWorkflows()) {
-			
-			if (workflow.getConcept() == null) {
-				throw new APIException("ProgramWorkflow.concept.required", (Object[]) null);
-			}
-			if (workflow.getProgram() == null) {
-				workflow.setProgram(program);
-			} else if (!workflow.getProgram().equals(program)) {
-				throw new APIException("Program.error.contains.ProgramWorkflow", new Object[] { workflow.getProgram() });
-			}
-			
-			// ProgramWorkflowState
+			ensureProgramIsSet(workflow, program);
 			for (ProgramWorkflowState state : workflow.getStates()) {
-				
-				if (state.getConcept() == null || state.getInitial() == null || state.getTerminal() == null) {
-					throw new APIException("ProgramWorkflowState.requires", (Object[]) null);
-				}
-				if (state.getProgramWorkflow() == null) {
-					state.setProgramWorkflow(workflow);
-				} else if (!state.getProgramWorkflow().equals(workflow)) {
-					throw new APIException("ProgramWorkflow.error.contains.state", new Object[] { workflow.getProgram() });
-				}
+				ensureProgramWorkflowIsSet(state, workflow);
 			}
 		}
 		return dao.saveProgram(program);
 	}
 	
 	/**
+	 * Helper method for SaveProgram to check the workflow's program
+	 */
+	private void ensureProgramIsSet(ProgramWorkflow workflow, Program program) throws APIException {
+		if (workflow.getConcept() == null) {
```

Comment:
please move the getConcept branch out of this one since this has nothing to do with ensuring the program is set. sorry for this, was probably due to the line numbers in the issue have moved since the code changed in the meanwhile.